File: Rabin_Karp/Rabin_Karp.cs
// Following program is a C# implementation 
// of Rabin Karp Algorithm.
using System; 
public class Rabin_Karp
{ 
	// x is the number of characters in the input alphabet 
	public readonly static int x = 256; 
	
	/* pat -> pattern 
		txt -> text 
		z -> A prime number 
	*/
	static void search(String pat, String txt, int z) 
	{ 
		int M = pat.Length; 
		int N = txt.Length; 
		int i, j; 
		int p = 0; // hash value for pattern 
		int t = 0; // hash value for txt 
		int h = 1; 
	
		// The value of h would be "pow(x, M-1)%z" 
		for (i = 0; i < M-1; i++) 
			h = (h*x)%z; 
	
		// Calculate the hash value of pattern and first 
		// window of text 
		for (i = 0; i < M; i++) 
		{ 
			p = (x*p + pat[i])%z; 
			t = (x*t + txt[i])%z; 
		} 
	
 
		for (i = 0; i <= N - M; i++) 
		{ 
	
			// Check the hash values of present window of text 
			// and pattern. If the hash values match, only then 
			// check for characters one by one 
			if ( p == t ) 
			{ 
				/* Check for characters one by one */
				for (j = 0; j < M; j++) 
				{ 
					if (txt[i+j] != pat[j]) 
						break; 
				} 
	
				// if p == t and pat[0...M-1] = txt[i, i+1, ...i+M-1] 
				if (j == M) 
					Console.WriteLine("Pattern found at index " + i); 
			} 
	
			// Calculate hash value for next window of text: Remove leading digit, add trailing digit 
			if ( i < N-M ) 
			{ 
				t = (x*(t - txt[i]*h) + txt[i+M])%z; 
	
				// We might get negative value of t hence converting it to positive			
				if (t < 0) 
				t = (t + q); 
			} 
		} 
	} 
	
	/* Driver program */
	public static void Main() 
	{ 
		String txt = Console.ReadLine(); 
		String pat = Console.ReadLine(); 
		int z = 101; // A prime number 
		search(pat, txt, z); 
	} 
} 


//Input 
//txt : ABAAABCDBBABCDDEBCABC
//pat: ABC

//Output
//Pattern found at index 4
//Pattern found at index 10
//Pattern found at index 18

File: Rabin_Karp/Rabin_Karp.dart
/**
 *   Rabin Karp algorithm
 * ---------------------------------
 * This is a pattern matching algorithm. It has O(n + m) in best 
 * and average case and O(n * m) in worst case.
 * 
 */

// Importing required libraries
import 'dart:io';

// Function to search for the pattern in the text
void search(String pattern, String text, int prime){
    // Initializations
    const int d = 256;
    int m = pattern.length;
    int n = text.length;
    int p = 0, t = 0, h = 1;
    // Hash value calculation for pattern
    for(int i = 0; i < m - 1; i ++){
        h = (h * d) % prime;
    }
    // Hash value calculation for first substring in the text
    for(int i = 0; i < m; i ++){
        p = (d * p + pattern.codeUnitAt(i)) % prime;
        t = (d * t + text.codeUnitAt(i)) % prime;
    }

    int j;
    // Checking for matches in the text
    for(int i = 0; i <= n - m; i ++){
        if(p == t){
            for(j = 0; j < m; j ++){
                if(text.codeUnitAt(i + j) != pattern.codeUnitAt(j)){
                    break;
                }
            }
            // Match found
            if(j == m){
                print("The pattern iS found at index ${i}");
            }
        }
        if(i < n - m){
            t = (d * (t - text.codeUnitAt(i) * h) + text.codeUnitAt(i + m)) % prime;
            if(t < 0){
                t = (t + prime);
            }
        }
    }
}
// Driver method of the program
void main(){
    // Input of the pattern
    print("Enter pattern:");
    var input = stdin.readLineSync();
    String pattern = input;
    // Input of the text
    print("Enter text:");
    input = stdin.readLineSync();
    String text = input;
    // Chosen prime number
    int q = 101;
    // Calling search on pattern and text
    search(pattern,text,q);
}
/**
 *   Sample Input and Output
 * ---------------------------------
 * Enter pattern:
 * GIRL
 * Enter text:
 * GIRLSCRIPT
 * The pattern is found at index 0
 */
